ironfish-cli/src/ui/ledger.ts

```diff
@@ -0,0 +1,62 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { PromiseUtils } from '@ironfish/sdk'
+import { ux } from '@oclif/core'
+import {
+  Ledger,
+  LedgerAppNotOpen,
+  LedgerClaNotSupportedError,
+  LedgerConnectError,
+  LedgerDeviceLockedError,
+  LedgerGPAuthFailed,
+  LedgerPortIsBusyError,
+} from '../ledger'
+
+export async function ledgerAction<TResult>(
+  ledger: Ledger,
+  action: () => TResult | Promise<TResult>,
+): Promise<TResult> {
+  const statusRunning = ux.action.running
+  let statusAdded = false
+
+  try {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      try {
+        const result = await ledger.tryInstruction(async () => await action())
+        ux.action.stop()
+        return result
+      } catch (e) {
+        if (e instanceof LedgerConnectError) {
+          ux.action.start('Ledger not found, connect ledger')
+        } else if (e instanceof LedgerDeviceLockedError) {
+          ux.action.start('Ledger is locked, enter pin code')
+        } else if (e instanceof LedgerPortIsBusyError) {
+          ux.action.start('Ledger port is already in use')
+        } else if (e instanceof LedgerGPAuthFailed) {
+          ux.action.start('Ledger handshake failed... trying again')
+        } else if (e instanceof LedgerClaNotSupportedError) {
+          const appName = ledger.isMultisig ? 'Ironfish' : 'Ironfish DKG'
+          ux.action.start(`Wrong Ledger app opened. Please open ${appName} APP.`)
+        } else if (e instanceof LedgerAppNotOpen) {
+          const appName = ledger.isMultisig ? 'Ironfish' : 'Ironfish DKG'
+          ux.action.start(`Unlock your Ledger and open the ${appName} APP.`)
+        } else {
+          throw e
+        }
```

Comment:
seeing this work is so sick. assuming once this PR gets in we need to use ledgerAction everywhere?